spec/models/incoming_message_spec.rb

```diff
@@ -115,7 +115,7 @@
                         :text => 'Person',
                         :info_request => message.info_request)
 
-      expect(message.safe_mail_from).to eq('FOI [REDACTED]')
+      expect(message.reload.safe_mail_from).to eq('FOI [REDACTED]')
```

Comment:
Would be good to include the failing spec output in the commit message, as its really difficult to understand _why_ this was failing.

Reply:
Have you looked at whether its safe to do this in terms of it working as we'd expect?

When calling `safe_mail_from` I _think_ I'd always expect to get a fresh cache of censor rules as we haven't directly loaded the associated info request or censor rules in the spec setup (though I haven't looked through the full method call chain in detail).

```diff
diff --git a/spec/models/incoming_message_spec.rb b/spec/models/incoming_message_spec.rb
index c47e6fa3f..cf4c66a9b 100644
--- a/spec/models/incoming_message_spec.rb
+++ b/spec/models/incoming_message_spec.rb
@@ -114,6 +114,7 @@ describe IncomingMessage do
       FactoryBot.create(:censor_rule,
                         :text => 'Person',
                         :info_request => message.info_request)
+      binding.pry
 
       expect(message.safe_mail_from).to eq('FOI [REDACTED]')
     end
```

```
# Rails 4
message.info_request.censor_rules
# => [#<CensorRule:0x00000010b08fc8
  id: 875,
  info_request_id: 112,
  user_id: nil,
  public_body_id: nil,
  text: "Person",
  replacement: "[REDACTED]",
  last_edit_editor: "FactoryBot",
  last_edit_comment: "Modified by rspec",
  created_at: Tue, 12 Feb 2019 16:21:29 UTC +00:00,
  updated_at: Tue, 12 Feb 2019 16:21:29 UTC +00:00,
  regexp: nil>]

# Rails 5
message.info_request.censor_rules
# => []
```

From the info in this commit message its not clear whether we've established that in real operation we always end up with the most recently added censor rules. Basically I'm wondering whether the `reload` should be in the app code itself, rather than the spec.

Reply:
Hmm... looks like it's actually `parse_raw_email!(true)` that causes the need to reload as this works:

```
      message = FactoryBot.create(:incoming_message)
      message.raw_email.data = raw_email_data
      message.parse_raw_email!(true)
      message.reload
      FactoryBot.create(:censor_rule,
                        :text => 'Person',
                        :info_request => message.info_request)

      expect(message.safe_mail_from).to eq('FOI [REDACTED]')
```

Reply:
which makes me wonder whether I need to do this instead...

```diff
diff --git a/app/models/incoming_message.rb b/app/models/incoming_message.rb
index 52112e9d1..849c631bf 100644
--- a/app/models/incoming_message.rb
+++ b/app/models/incoming_message.rb
@@ -110,6 +110,7 @@ class IncomingMessage < ActiveRecord::Base
         self.foi_attachments.reload
         self.save!
       end
+      self.info_request.reload
     end
   end
```

Reply:
Ah - found it, fixed in `parse_raw_email!` saves and a callback to update the related request inadvertently caches it before we add the censor rule. Fixed in 116e80907

Reply:
@lizconlan from my understanding your analysis looks correct and what you've done in 4d02788, reloading before the censor rule is created, looks to be the simplest solution to this.